d2/src/test/java/com/linkedin/d2/balancer/servers/ZookeeperConnectionManagerTest.java

```diff
@@ -471,6 +472,311 @@ public void testMarkUpAndDownMultipleTimesFinalUp()
     executorService.shutdown();
   }
 
+  @Test
+  public void testNoWarmupWhenDisabled() throws Exception
+  {
+    ScheduledExecutorService warmupExecutorService = Executors.newSingleThreadScheduledExecutor();
+    boolean isDarkWarmupEnabled = false;
+    String warmupClusterName = "warmup" + _cluster;
+    int warmupDuration = 5; //run warm up for 5 sec
+
+    ZooKeeperAnnouncer announcer = getZooKeeperWarmupAnnouncer(_cluster, _uri, WEIGHT, isDarkWarmupEnabled, warmupClusterName, warmupDuration, warmupExecutorService);
+    ZooKeeperConnectionManager manager = createManagerForWarmupTests(false, warmupDuration, announcer);
+    ZooKeeperEphemeralStore<UriProperties> store = createAndStartUriStore();
+    FutureCallback<None> managerStartCallback = new FutureCallback<>();
+    manager.start(managerStartCallback);
+
+    //ugly but to ensure that manager has started and has announcer has announced the regular cluster
+    Thread.sleep(1000);
```

Comment:
Heavy +1 on this. Please avoid using Thread.sleep if possible, so that running all the tests won't be cumbersome. Please see the example provided by Lester on using a ClockedExecutor

Reply:
I couldn't get the ClockedExecutor to work. The schedule method of the Clocked Executor  (https://github.com/rpuranda/rest.li/blob/master/test-util/src/main/java/com/linkedin/test/util/ClockedExecutor.java#L94) wasn't working to schedule the markDown for the warm-up cluster (https://github.com/linkedin/rest.li/pull/726/files#diff-ce9b6eecd634167a616ca98580fd1b888067137f94c92c2e92026dc182f946ebR361). I have changed the test cases in a way to get the callback of ZooKeeperConnectionManager after a timeout to ensure warm-up has begun, to avoid using Thread.sleep. There is only 1 test case now that needs using a Thread.sleep after zkServer restart, to check that zkServer has started. Please review if this is acceptable.